lib/appsignal/rack/hanami_instrumentation.rb

```diff
@@ -0,0 +1,54 @@
+require "rack"
+
+module Appsignal
+  # @api private
+  module Rack
+    class HanamiInstrumentation
+      def initialize(app, options = {})
+        Appsignal.logger.debug "Initializing Appsignal::Rack::HanamiInstrumentation"
+        @app = app
+        @options = options
+      end
+
+      def call(env)
+        if Appsignal.active?
+          call_with_appsignal_monitoring(env)
+        else
+          @app.call(env)
+        end
+      end
+
+      def call_with_appsignal_monitoring(env)
+        request = ::Rack::Request.new(env)
+        transaction = Appsignal::Transaction.create(
+          SecureRandom.uuid,
+          Appsignal::Transaction::HTTP_REQUEST,
+          request
+        )
+        begin
+          Appsignal.instrument("process_action.generic") do
+            @app.call(env)
+          end
+        rescue => error
+          transaction.set_error(error)
+          raise error
+        ensure
+          controller = hanami_action(env)
+          if !controller.empty?
+            transaction.set_action_if_nil(controller)
+          else
+            transaction.set_action_if_nil("unknown")
+          end
+          transaction.set_metadata("path", request.path)
+          transaction.set_metadata("method", request.request_method)
+          transaction.set_http_or_background_queue_start
+          Appsignal::Transaction.complete_current!
+        end
+      end
+
+      def hanami_action(env)
+        Web.routes.recognize(env["REQUEST_PATH"]).action.to_s
```

Comment:
My Hanami is a bit rusty, but is this a constant something which can be customized by a developer? In case of multiple apps can there be more constant names?